File: 2408-number-of-people-aware-of-a-secret/2408-number-of-people-aware-of-a-secret.go
func peopleAwareOfSecret(n int, delay int, forget int) int {
    dp := make([]int, n + 1)
    for i := 0; i < n + 1; i++ {
        dp[i] = -1
    }
    mod := int(1e9 + 7)
    var helper func(int) int
    helper = func(i int) int {
        if i > n {
            return 0
        }
        if dp[i] != -1 {
            return dp[i]
        }
        sum := 1        // The person themself
        for k := delay ; k < forget && i + k <= n; k++ {
            sum = (sum + helper(k + i)) % mod
        }
        dp[i] = sum
        if i + forget <= n {
            dp[i]--
        }
        return dp[i]
    } 
    return helper(1)
}

File: 2408-number-of-people-aware-of-a-secret/2408-number-of-people-aware-of-a-secret.cpp
class Solution {
public:
    int peopleAwareOfSecret(int n, int delay, int forget) {
        int mod = 1e9 + 7;
        vector<int> dp(n + 1, -1);
        function<int(int)> helper = [&](int i) -> int {
            if(i > n) return 0;
            if(i == n) return 1;
            if(dp[i] != -1) return dp[i];
            long long sum = 0LL;
            for(int k = i + delay; k < i + forget; k++)
                sum = (sum + helper(k)) % mod;
            return dp[i] = static_cast<int>((i + forget) > n ? (sum + 1) % mod : sum);
        };
        return helper(1);
    }
};